semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -0,0 +1,180 @@
+
+  bool b = parens (str (if b then "true" else "false"))
+  string = parens . text
+
+  throw e = parens (str "throw" <+> e)
+
+  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
+
+str :: String -> Print
+str = Print . showString
+
+text :: Text -> Print
+text = str . unpack
+
+char :: Char -> Print
+char = Print . showChar
+
+parens :: Print -> Print
+parens p = char '(' <> p <> char ')'
+
+(<+>) :: Print -> Print -> Print
+l <+> r = l <> char ' ' <> r
+
+infixr 6 <+>
+
+name :: Name -> Print
+name = text . formatName
+
+
+-- Abstract interpretation
+
+eval0 :: Interpret m i -> m i
+eval0 = fix eval
+
+eval :: (Interpret m i -> m i) -> (Interpret m i -> m i)
+eval eval (Interpret f) = f eval
+
+newtype Interpret m i = Interpret { interpret :: (Interpret m i -> m i) -> m i }
+
+instance (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m) => Syntax (Interpret m val) where
+  iff c t e = Interpret (\ eval -> do
+    c' <- eval c
+    dif c' (eval t) (eval e))
+  noop = Interpret (const dunit)
+
+  bool b = Interpret (\ _ -> dbool b)
+  string s = Interpret (\ _ -> dstring s)
+
+  throw e = Interpret (\ eval -> eval e >>= ddie)
+
+  let_ n v b = Interpret (\ eval -> do
+    v' <- eval v
+    let' n v' (eval (b (Interpret (pure (pure v'))))))
+
+
+-- Macro-expressible syntax
+
+let' :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m) => Name -> val -> m a -> m a
+let' n v m = do
+  addr <- alloc n
+  addr .= v
+  bind n addr m
+
+
+-- Parsing
+
+parseFile :: Syntax rep => FilePath -> IO (Either (A.JSONPath, String) (Maybe rep))
+parseFile path = do
+  contents <- B.readFile path
+  pure $ snd <$> A.eitherDecodeWith A.json' (A.iparse parseGraph) contents
+
+parseGraph :: Syntax rep => A.Value -> A.Parser (IntMap.IntMap rep, Maybe rep)
+parseGraph = A.withArray "nodes" $ \ nodes -> do
+  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
+  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
+  pure (tied, ($ tied) <$> root)
+
+parseNode :: Syntax rep => A.Object -> A.Parser (IntMap.Key, IntMap.IntMap rep -> rep, Maybe (IntMap.IntMap rep -> rep))
+parseNode o = do
+  edges <- o A..: pack "edges"
+  index <- o A..: pack "id"
+  let parseType attrs = \case
+        "string" -> const . string <$> attrs A..: pack "text"
+        "true"   -> pure (const (bool True))
+        "false"  -> pure (const (bool False))
+        "throw"  -> fmap throw <$> resolve (head edges)
+        "if"     -> liftA3 iff <$> findEdgeNamed "condition" <*> findEdgeNamed "consequence" <*> findEdgeNamed "alternative" <|> pure (const noop)
+        "block"  -> children
+        "module" -> children
+        t        -> A.parseFail ("unrecognized type: " <> t)
+      children = fmap (foldr (\ (i, v) r -> let_ (nameI i) v (const r)) noop . zip [0..]) . sequenceA <$> traverse resolve edges
```

Comment:
This right here is pretty opaque; can you add a type signature and perhaps make it a little more verbose? Maybe pull out the fold function into something explicitly named?